### PR TITLE
feat: Add cli for simple gateway registration

### PR DESCRIPTION
**Pull request by christinewang5**
## Summary
register CLI for gateways

## Testing
manually tested by running CLIs 
image

Note: this PR is dependent on #11169 & #11692

### CODE REVIEW

File: orc8r/cloud/docker/controller/supervisord.conf
Comment (hcgatewood) [inline orc8r/cloud/docker/controller/supervisord.conf]: Nit: revert these 2

File: orc8r/cloud/docker/docker-compose.yml
Comment (hcgatewood) [inline orc8r/cloud/docker/docker-compose.yml]: This should probably be `magma.test`

File: orc8r/cloud/go/service/middleware/unary/identity_decorator_bypass.go
Comment (hcgatewood) [inline orc8r/cloud/go/service/middleware/unary/identity_decorator_bypass.go]: Shouldn't need to add this here -- "CloudRegistration" is called by "Registration", rather than directly from gw

Screen Shot 2022-02-23 at 2 32 47 PM

File: orc8r/cloud/go/services/bootstrapper/bootstrapper/main.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/bootstrapper/bootstrapper/main.go]: Let's just make this an err, rather than fatal

File: orc8r/gateway/python/magma/common/service_registry.py
Comment (hcgatewood) [inline orc8r/gateway/python/magma/common/service_registry.py]:
```suggestion
        Return an RPC channel to bootstrap service in CLOUD, without referencing a
        control proxy config.
```

File: orc8r/gateway/python/magma/common/service_registry.py
Comment (hcgatewood) [inline orc8r/gateway/python/magma/common/service_registry.py]: If we're gonna have this only target bootstrapper, then should reflect that in the func name

File: orc8r/gateway/python/magma/common/service_registry.py
Comment (hcgatewood) [inline orc8r/gateway/python/magma/common/service_registry.py]: The `magma.test` should be passed in as an arg

File: orc8r/gateway/python/magma/common/service_registry.py
Comment (hcgatewood) [inline orc8r/gateway/python/magma/common/service_registry.py]: Same thing for port

File: orc8r/gateway/python/magma/common/service_registry.py
Comment (hcgatewood) [inline orc8r/gateway/python/magma/common/service_registry.py]: Same

File: orc8r/gateway/python/scripts/generate_nghttpx_config.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/generate_nghttpx_config.py]: Revert before merging

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py]: Fmu why the comments?

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py]: Maybe like "bare_cloud_grpc_wrapper"

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py]: Match to func name

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py]: Can you create this as a literal rather than procedurally? E.g. how you construct a literal struct in golang

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py]: Use one big formatted string e.g. https://github.com/magma/magma/blob/33cd34820810ed71947aca9a6011182496c71324/orc8r/gateway/python/scripts/checkin_cli.py#L146-L152

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py]: As noted, this isn't necessary

And then, because we are only making a single request to the cloud, we may not need to use the "bare_grpc_wrapper" that you created -- just make the request manually, e.g. https://github.com/magma/magma/blob/33cd34820810ed71947aca9a6011182496c71324/orc8r/gateway/python/scripts/checkin_cli.py#L175-L194

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py]: I think we can do this without resorting to shell cmds. Maybe something like https://stackoverflow.com/questions/12517451/automatically-creating-directories-with-file-output

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Woo looks great, couple small things but happy we got it working!

File: orc8r/cloud/docker/docker-compose.yml
Comment (christinewang5) [reply orc8r/cloud/docker/docker-compose.yml]: Idk why I rebased this PR onto the domain name PR, but I just undid the rebase and fixed all these in the original domain name PR.

File: orc8r/gateway/python/scripts/register.py
Comment (christinewang5) [reply orc8r/gateway/python/scripts/register.py]: Yeah so I was originally doing that, but because we are trying to create a file from root, it threw a `Permission Denied`. 
I think it needs sudo permissions to make a dir in the `/var` folder, was also trying to chmod and setting umask, neither worked.

https://stackoverflow.com/questions/5231901/permission-problems-when-creating-a-dir-with-os-makedirs-in-python

File: orc8r/gateway/python/scripts/register.py
Comment (christinewang5) [inline orc8r/gateway/python/scripts/register.py]: not sure if there is a better way to tell whether we are in test env or not??

File: orc8r/gateway/python/scripts/register.py
Comment (christinewang5) [reply orc8r/gateway/python/scripts/register.py]: Hey @ardzoht! 
Was wondering if there are better ways to grant write access to a file from the AGW root. Are there unintended side effects of doing a chmod 777 like this? Thanks in advance!

Comment (ardzoht) [reply orc8r/gateway/python/scripts/register.py]: @christinewang5 I'd recommend to look up for the function handlers on `service_registry`, I think we can make use of these that should accomplish the same thing: https://sourcegraph.com/github.com/magma/magma/-/blob/orc8r/gateway/python/magma/configuration/service_configs.py?L47:5

The only caveat is that this probably will require running the script with sudo permissions, but it should work and the code is cleaner.

Comment (christinewang5) [reply orc8r/gateway/python/scripts/register.py]: Thanks for the pointer! @ardzoht 
Could you please advise on how to run the script with sudo permissions in the magtivate virtualenv? 
When I run it, I have to put the whole abs path there like (`sudo /home/vagrant/magma/orc8r/gateway/python/scripts/register.py`) and even then it throws a cannot find the module snowflake.

I tried several things like running with the -E option [1] and also changing /etc/sudoers which seems sketch [2] without avail, but it seems like running with sudo is using a diff python environment. 
[1] [https://stackoverflow.com/questions/50315645/cannot-run-python-script-using-sudo](https://stackoverflow.com/questions/50315645/cannot-run-python-script-using-sudo)
[2] [https://stackoverflow.com/questions/25346171/cant-import-module-when-using-root-user](https://stackoverflow.com/questions/25346171/cant-import-module-when-using-root-user)

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [reply orc8r/gateway/python/scripts/register.py]: I'd say let's make it a CLI arg, default to 8444. Then in test env you have to pass e.g. `--cloud-port 7444`

File: orc8r/gateway/python/magma/common/service_registry.py
Comment (ardzoht) [inline orc8r/gateway/python/magma/common/service_registry.py]: can we define the filepath as a constant?

File: orc8r/gateway/python/magma/common/service_registry.py
Comment (ardzoht) [inline orc8r/gateway/python/magma/common/service_registry.py:105]: is `authority` the same as `ip`?

Comment (christinewang5) [reply orc8r/gateway/python/magma/common/service_registry.py:105]: in this case, yes, will just use `authority` for uniformity

**Review by ardzoht — APPROVED** [review]
lgtm

File: orc8r/gateway/python/magma/common/service_registry.py
Comment (hcgatewood) [inline orc8r/gateway/python/magma/common/service_registry.py]: I believe you can do

```suggestion
        rootca_path = rootca_path or ROOTCA_PATH
```

Comment (hcgatewood) [inline orc8r/gateway/python/magma/common/service_registry.py]: So if no rootca path is passed in, we want to use the default system credentials. I *think* that means you don't have to pass any ssl creds, but not sure how it works. But if you want, we can leave it like this for now, and I'll test it when I get around to making the Orc8r<>Let'sEncrypt work e2e

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py]: Your args are mixing dash-case and camel_case -- let's use dash-case since it's historically the default for CLIs

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py:76]: Is there a way to mark these arguments as required (and fail/throw err when they're not present) ?

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py:123]: Final thing we wanna do here is run the "checkin_cli.py" functionality, so that the operator can have confidence that the registration was fully successful

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Looking good! Couple small things

File: orc8r/gateway/python/scripts/register.py
Comment (christinewang5) [reply orc8r/gateway/python/scripts/register.py:76]: Soo because these are considered `positional arguments` , python's `argparse` already throws an error when these are not present

Comment (hcgatewood) [reply orc8r/gateway/python/scripts/register.py:76]: Oh okay cool. And the "--xyz" args are magically non-positional, then? Nice

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py]: Nit: I'd prob put this printed msg in the main func, to keep this func as "stateless" as possible

File: orc8r/gateway/python/scripts/register.py
Comment (hcgatewood) [inline orc8r/gateway/python/scripts/register.py:128]: Oh dang does this work? Nice good to know

**Review by hcgatewood — APPROVED** [review]
Lgtm!

File: orc8r/gateway/python/scripts/register.py
Comment (christinewang5) [reply orc8r/gateway/python/scripts/register.py:128]: Yeah looks a bit sketch but it does 🤷🏻‍♀️

**Pull request merged**
